### DISCUSSION TITLE
Grafana "query data error" after following docker installation guide

### DISCUSSION THREAD

**Discussion by Bz16** [teslamate-org/teslamate#2773] [Q&A] [answered]
Hi,

I just installed the TeslaMate 1.27.0 on my Synology DS420+ via Docker (v20.10.3, apparently) two days ago. I used the generic docker-compose file from the [docker installation guide](https://docs.teslamate.org/docs/installation/docker), though I modified the whole setup a bit later manually:

1. I renamed the containers and changed all `*_HOST` environment variables to match, 
2. Let Docker mapped TeslaMate's port 4000 to ext. port 6394, and Grafana's port 3000 to ext. port 6393 (I just wanted to keep my exposed self-use port tidy), 
3. Mapped all volumes mapped in docker-compose into an easily accessible one, e.g. `/volume2/docker/teslamate/import:/opt/app/import` (Synology's Docker implementation uses a hidden directory as its working directory, so I followed common Synology practice and used an accessible share folder for all these), and
4. Setup the bridge network myself, so it will have a sensible name.

So far, I am able to access TeslaMate's and Grafana's web interface, and TeslaMate seems to be writing into the database correctly – I tried looking into the postgres database and there seems to be some data; however, Grafana keeps throwing "query data errors" for everything, even though it should able to connect and access the database. 

This is the returned error:
>status:500
>statusText:"Internal Server Error"
>data:Object
>  message:"Query data error"
>  traceID:"00000000000000000000000000000000"
>config:Object
>  url:"api/ds/query"
>  method:"POST"
>data:Object
>  requestId:"Q103"
>  retry:0
>  headers:Object
>  hideFromInspector:false
>message:"Query data error"

The docker-compose logs shows Grafana repeatedly making complaints like these (apparently whenever I ask it to query the database):
>TM-grafana   | logger=tsdb.postgres t=2022-07-18T19:11:45.52+0000 lvl=eror msg="Failed connecting to Postgres" err="invalid option: \"RXtCmuEo6MbKd6NUKrjWMkMd'\""
>TM-grafana   | logger=context traceID=00000000000000000000000000000000 userId=1 orgId=1 uname=admin t=2022-07-18T19:11:45.53+0000 lvl=eror msg="Query data error" error="failed to query data: invalid option: \"RXtCmuEo6MbKd6NUKrjWMkMd'\"" remote_addr=172.19.0.1 traceID=00000000000000000000000000000000

Looks like Postgres never complained; though today there's been a flood of API timeout errors and stale data from teslamate, but this happened after the Grafana issue and should be unrelated.

Quick nmap test from an Alpine-Linux container in the same bridge network shows that all containers' host name works, and all containers other then mosquitto have open ports as required (4000, 3000, 5432). I don't know how to send a Postgres query from command line, so I haven't try that.

Anything else should I provide? Any recommendations as to what to next?

Thanks for any help in advance.

**Comment by VinciSC** [comment]
Hi,
same issue here on a QNAP TS251+. Recently installed teslamate (latest) in QNAPs container station according to [docker installation guide](https://docs.teslamate.org/docs/installation/docker). No local modifications on the yml file except database pw and encryption key. 
Web interfaces of teslamate and grafana are accessible, teslamate shows correct current position of vehicle but grafana shows exclamation mark and "Data query error" in all dashboards.
No idea how to check "docker-compose logs" as I'm completely new in docker and QNAPs container station.
Any recommendation appreciated, I'll try out every hint.
Thanks and best regards
VinciSC

**Reply by Bz16** [reply]
The docker-compose logs can be accessed in the terminal: `cd` to the Docker directory – where you placed the docker-compose.yml file – just like how you used the Docker installation guide to setup the containers; but, instead of running `docker-compose up`, run `docker-compose logs`.

You can also use `docker-compose logs | grep 'grafana'` to show only logs from/related to grafana (assuming your grafana container is indeed named `grafana`); replace `grafana` with any other container's name for logs from/related to another container. 

Maybe you can check your Grafana's and Postgres's logs to see if our issues are similar? Your symptom is basically identical as mine: a working TeslaMate web portal and accessible but non-functional Grafana web portal due to "Data query errors".

**Reply by VinciSC** [reply]
Was a bit fiddely for me to find the directory because I used the graphical interface of QNAPs container station for setup. Finally I was able to list the grafana log and found several (better: a lot of) entries identical to yours: 

> grafana_1    | logger=tsdb.postgres t=2022-07-20T07:23:34.56+0000 lvl=eror msg="Failed connecting to Postgres" err="invalid option: \"xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx'\""
> grafana_1    | logger=context traceID=00000000000000000000000000000000 userId=1 orgId=1 uname=admin t=2022-07-20T07:23:34.56+0000 lvl=eror msg="Query data error" error="failed to query data: invalid option: \"xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx'\"" remote_addr=192.168.178.20 traceID=00000000000000000000000000000000

The string I replaced with xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx is my database pw. Confusing that it reports "invalid option" followed by a pw!?

**Reply by VinciSC** [reply]
Addendum: When clicking on the excalmation mark in a grafana dashboard I get the following error message in the inspect window:
> Query data error
> Object
> status:500
> statusText:"Internal Server Error"
> data:Object
> message:"Query data error"
> traceID:"00000000000000000000000000000000"
> config:Object
> url:"api/ds/query"
> method:"POST"
> data:Object
> requestId:"Q113"
> retry:0
> headers:Object
> hideFromInspector:false
> message:"Query data error"

**Reply by Bz16** [reply]
> The string I replaced with xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx is my database pw. Confusing that it reports "invalid option" followed by a pw!?

Whoa, I totally forgot to check mine – double-checked and that is my password there too. Oh well, my TeslaMate instance won't be exposed to the internet, so I should be fine for now...

Can you check your docker-compose.yml and see if you filled in your database password like so:
`POSTGRES_PASSWORD= RXtCmuEo6MbKd6NUKrjWMkMd`,
with a space between the equal sign and the string? I accidentally did that, and it was treated as part of the password. Looks like the space was stripped in Grafana query, and that caused all the issue.

**Reply by VinciSC** [reply]
Yes, I had a space between equal sign and the pw string at all 3 pw environment variables. Can't relocate at the moment but I read a post in a Tesla web forum to enter the pw in that way. Was wondering about this deviation from all other env variables but didn't scrutinize. 
Changed the pw and updated the yml file now and it seems to work so far, access to grafana dashboards doesn't throw errors any more. Have to drive some km to see if and how it grabs data ...
Thanks for supporting me and best regards!
Vincent

**Comment by Bz16** [comment, marked as answer]
So it looks like the reason is that when I am following the docker-compose.yml file, wherever there is things like
`- DATABASE_PASS= #insert your secure database password!`,
I replaced only `#` and the string after `#` to my password of choice, like so:
`- DATABASE_PASS= my-password-is-very-safe`
Thus the password environmental variables all include a prefixed space. Looks like Grafana strips the space when sending the query and that causes all the errors.

**To Fix,** just change the password to one that have no space:
1. Stop all containers other then Postgres Database.
2. Change the Postgres password. For us NAS users with GUI, enter the Postgres terminal, and run
- `psql -U teslamate -d teslamate`
- `\password teslamate`
-  and follow the prompt – make sure to not use spaces.
3. Stop Postgres. Go through Postgres, Grafana, and TeslaMate, and change all password environmental variables to the new password.
4. Restart all containers.

**To Avoid,** just make sure you delete the space before `#` when setting up docker-compose.yml.

**Reply by jferguson6** [reply]
Thank you, I was beating my head trying to figure this exact problem....

**Comment by WICKII** [comment]
Though @Bz16 has given the answer, would like to share a quick way to debug this issue.
Just check if user name and pwd is matching with docker-compose.yml in Grafana.
There is quick way to "save and test" to see if two are matching.
![image](https://github.com/user-attachments/assets/24fba3f0-5172-45e9-8e3d-bdc3e1c2ce76)

**Comment by AlDoddsie** [comment]
Tore a bit of hair out chasing this issue to find that the answer (in my case) was simple. The docker-compose.yml file calls the database "database". The grafana page at url:3000, under Connections->Data sources (as above) showed the Host URL as db:5432. Changing this to database:5432, solved the issue and all my data popped up. Phew!
